### ISSUE TITLE
Theme support

### ISSUE THREAD

**Issue by ws909** [glfw/glfw#2257]
This is a feature request for adding basic theming support to GLFW.

Goals:
- Query the current (default) system theme (light/dark).
- Query for a system theme colour if the platform supports it.
- Get notified every time the system changes its theme (user edited it, MacOS's automatic day/night, etc).
- Set a window's theme to change automatically based on the system theme (default GLFW behaviour).
- Set a window's theme (light/dark).
- Set a specific theme colour for a window, if the platform supports it.

Similar issues and PRs:
- #1516
- #2228

This is a very rough idea of what this API might look like:

```c
typedef struct GLFWtheme
{
    int baseTheme; // light/dark
    int hasColor;
    float color[4];
} GLFWtheme;

#define GLFW_THEME_LIGHT 0
#define GLFW_THEME_DARK 1

typedef void (* GLFWthemefun)(GLFWtheme* theme);

/*! @brief Notifies the application when the system default theme changes.
 *
 */
GLFWAPI void glfwSetSystemThemeCallback(GLFWthemefun callback);

/*! @brief Sets the theme for a window.
 *
 *  @param[in] window The window to set the theme for.
 *  @param[in] theme The theme to set. Pass `NULL` to set it to the system default.
 */
GLFWAPI void glfwSetTheme(GLFWwindow* window, GLFWtheme* theme);
GLFWAPI GLFWtheme* glfwGetTheme(GLFWwindow* window);
GLFWAPI GLFWtheme* glfwGetSystemDefaultTheme();
```

Windows supports custom colours for its window borders and title bars. The libdecor decorations for Wayland likely supports this too. MacOS does not support custom colours, but does support light/dark modes. MacOS does however have an accent colour, which is used for text. The accent colour is both gettable and settable.

**Comment by elmindreda** [comment]
I agree with all of these goals. Thank you for making an issue collecting them! Heading to the PR for comments on the proposed API.

**Comment by ws909** [comment]
@elmindreda I just found a branch of yours from 4 years ago, and it seems like you did some [work](https://github.com/elmindreda/glfw/commit/268d699bfc40b2d3408b168e479d593e9f79223b) on getting theming into GLFW. You only implemented dark/light mode querying on Cocoa. May I ask what the status of the other platforms is?

**Comment by elmindreda** [comment]
Oh, yup, I poked the problem a tiny bit but I have no code beyond what you found and I kept that mostly as a reminder. The combination of needing terrible hacks on every other platform and other features having higher priority made me put it aside.

**Comment by ws909** [comment]
@elmindreda I have no experience with the other platforms, but I don’t want my PR to linger to death, so if too much time passes without anyone helping with the implementations for these platforms, I may get a VM and have a look at this myself, so if you have any knowledge on how this must be implemented on these platforms, that would be great. It would also help with designing the API. For instance, the current «vibrancy» attribute is very MacOS specific, and should probably be removed, but the more knowledge I have of the other platforms, the better. And if I do implement this myself, it would be great to have some basis to go off from. It doesn’t seem so far as if WL or X11 support theming, but I assume linking to dynamic libraries, and reading system properties, are fine. So if you could share what you figured out from your research back then, that would be great.

**Comment by elmindreda** [comment]
@ws909 Sorry, lost all usable time for a bit. I'll leave feedback as soon as I'm able.

**Comment by ws909** [comment]
@elmindreda The absence of any work on this, has led me to realize that the API I designed isn't very good. It doesn't provide adequate means of checking if a window's theme properties are inherited from the system theme, if the system has these properties at all, or if they've been set by the user. So, in short, don't prioritize looking at my PR.

**Comment by MirrasHue** [comment]
SDL has implemented this I guess, could it be used as a reference?
